pkg/app/piped/imageprovider/ecr/ecr.go

```diff
@@ -0,0 +1,244 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ecr
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ecr"
+	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+// The maximum number of image results returned by the APIs about images.
+// This value can be between 1 and 1000.
+// See more: https://pkg.go.dev/github.com/aws/aws-sdk-go/service/ecr#ListImagesInput
+const maxResults = 1000
+
+type ECR struct {
+	name            string
+	client          *ecr.ECR
+	region          string
+	credentialsFile string
+	profile         string
+	registryID      string
+
+	logger *zap.Logger
+}
+
+type Option func(*ECR)
+
+func WithRegistryID(id string) Option {
+	return func(e *ECR) {
+		e.registryID = id
+	}
+}
+
+func WithCredentialsFile(path string) Option {
+	return func(e *ECR) {
+		e.credentialsFile = path
+	}
+}
+
+func WithProfile(profile string) Option {
+	return func(e *ECR) {
+		e.profile = profile
+	}
+}
+
+func WithLogger(logger *zap.Logger) Option {
+	return func(e *ECR) {
+		e.logger = logger
+	}
+}
+
+func NewECR(name string, region string, opts ...Option) (*ECR, error) {
+	if region != "" {
+		return nil, fmt.Errorf("region is required")
+	}
+	e := &ECR{
+		name:   name,
+		logger: zap.NewNop(),
+	}
+	for _, opt := range opts {
+		opt(e)
+	}
+	e.logger = e.logger.Named("ecr-provider")
+
+	cfg := aws.NewConfig().WithRegion(e.region)
+	if e.credentialsFile != "" {
+		cfg = cfg.WithCredentials(credentials.NewSharedCredentials(e.credentialsFile, e.profile))
+	} else {
+		cfg = cfg.WithCredentials(credentials.NewEnvCredentials())
+	}
+	sess := session.Must(session.NewSession())
+	e.client = ecr.New(sess, cfg)
+	return e, nil
+}
+
+func (e *ECR) Name() string {
+	return e.name
+}
+
+func (e *ECR) Type() model.ImageProviderType {
+	return model.ImageProviderTypeECR
+}
+
+func (e *ECR) ParseImage(image string) (*model.ImageName, error) {
+	ss := strings.SplitN(image, "/", 2)
+	if len(ss) < 2 {
+		return nil, fmt.Errorf("invalid image format (e.g. account-id.dkr.ecr.region.amazon.aws.com/pipecd/helloworld)")
+	}
+	return &model.ImageName{
+		Domain: ss[0],
+		Repo:   ss[1],
+	}, nil
+}
+
+func (e *ECR) GetLatestImage(ctx context.Context, image *model.ImageName) (*model.ImageRef, error) {
+	input := &ecr.ListImagesInput{
+		RepositoryName: aws.String(image.Repo),
+		Filter:         &ecr.ListImagesFilter{TagStatus: aws.String("TAGGED")},
+		MaxResults:     aws.Int64(maxResults),
+	}
+	if e.registryID != "" {
+		input.RegistryId = &e.registryID
+	}
+
+	// TODO: Consider the way to determine the latest tag other than fetching all tags
```

Comment:
Or instead of checking by version, caching the image metadata in the memory cache to reduce the API calls.
(Not sure our clients follow the convention of semantic versioning.)

Reply:
Do they have any options for sorting the results? I think no. lol.

Reply:
>Do they have any options for returning the sorted data by the created date? I think no. lol.

As you assumed, it doesn't ensure it, haha.

>caching the image metadata in the memory cache

Sorry if this sounds dumb, does it mean caching the image digest and compare the latest image's, then start fetching all tags if there is any deviation?